wsgi_kerberos.py

```diff
@@ -40,25 +40,39 @@ def _consume_request(environ):
     defending itself against denial-of-service attacks, or from badly broken
     client implementations.
     '''
+    if environ.get("HTTP_EXPECT") == "100-continue":
+        return
     try:
         sock = environ.get('wsgi.input')
         if hasattr(sock, 'closed') and sock.closed:
             return
-        # Figure out how much content is available for us to consume.
-        expected = int(environ.get('CONTENT_LENGTH', '0'))
+
+        content_length = environ.get('CONTENT_LENGTH')
+        if content_length is None:
+            remaining = read_max
+        else:
+            content_length = int(content_length)
+            if content_length > read_max:
+                # We are not willing to read such a large request body, but
+                # reading anything less does not help naively-written clients.
+                return
+            remaining = content_length
 
         # Try to receive all of the data. Keep retrying until we get an error
         # which indicates that we can't retry. Eat errors. The client will just
         # have to deal with a possible Broken Pipe -- we tried.
-        received = 0
-        while received < expected:
+        while remaining > 0:
             try:
-                received += len(sock.read(expected - received))
+                delta = len(sock.read(remaining))
             except socket.error as err:
                 if err.errno != errno.EAGAIN:
                     break
-    except (KeyError, ValueError):
```

Comment:
(KeyError looks like it was mistakenly left in here after e.g. `environ['CONTENT_LENGTH']` was changed to `environ.get('CONTENT_LENGTH')` above in an earlier revision? And ValueError looks like it's trying to account for the `int(content_length_string)` conversion, but I think WSGI servers should already validate that Content-Length headers contain valid integers before we get here?)

It looks like this should just catch Exception in any case -- this is a best effort function that should not ever cause an unhandled exception and resulting 500 error on the server. I made that improvement below, and also added a debug log statement with the exception info so the user actually has a way of finding out more about any swallowed exceptions.

Reply:
That's fine, though I'm not sure about printing the stack trace (even in debug mode) - it might confuse developer into thinking there is an actual problem when there isn't. Should be fine to just print just the message. I'd also be explicitly about the fact that the exception comes from wsgi_kerberos so devs don't have to chase that down, eg: `wsgi_kerberos:_consume_request failed: str(exc)`

It might actually be better to just print nothing at all (as before). Unless you're a wsgi_kerberos developer, I can't imagine why devs would care about seeing debug logs of clients sending invalid content-length headers.

Reply:
> I think WSGI servers should already validate that Content-Length headers contain valid integers before we get here

I wouldn't rely on that - at least the wsgiref server doesn't do any validation.